File: 239-SlidingWindowMaximum/239-SlidingWindowMaximum.py
# Last updated: 8/13/2025, 8:20:39 PM
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue = deque()
        result = []
        for i in range(k) : 
            while len(queue) >0 and queue[-1] < nums[i] : 
                queue.pop()
            
            queue.append(nums[i])
        
        result.append(queue[0])

        start = 0
        for i in range(k , len(nums)) : 
            while len(queue) >0 and queue[-1] < nums[i] : 
                queue.pop()
            
            queue.append(nums[i])

            if queue[0] == nums[i-k] : 
                queue.popleft()
            
            result.append(queue[0])
        
        return result

File: 239-SlidingWindowMaximum/239-SlidingWindowMaximum.java
// Last updated: 8/13/2025, 8:20:45 PM
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> q = new ArrayDeque();
        int n = nums.length;
        int ans[] = new int[n-k+1];

        for(int i=0;i<k;i++){
            while(q.size()>0 && nums[i]>q.getLast()){
                q.removeLast();
            }
            q.addLast(nums[i]);
        }
        ans[0] = q.getFirst();

        for(int i=k;i<n;i++){
            if(nums[i-k]==q.getFirst()){
                q.removeFirst();
            }
            while(q.size()>0 && nums[i]>q.getLast()){
                q.removeLast();
            }
            q.addLast(nums[i]);
            ans[i-k+1] = q.getFirst();
        }

        return ans;
    }
}